File: backend/db/migrate.ts
import * as fs from 'fs';
import * as path from 'path';
import { Pool } from 'pg';

const pool = new Pool({
  host: process.env.PGHOST ?? 'localhost',
  port: parseInt(process.env.PGPORT ?? '5433', 10),
  database: process.env.PGDATABASE ?? 'axiom_dev',
  user: process.env.PGUSER ?? 'axiom',
  password: process.env.PGPASSWORD ?? 'axiom',
});

async function migrate(): Promise<void> {
  // Bootstrap: ensure schema_migrations exists before we can track anything
  await pool.query(`
    CREATE TABLE IF NOT EXISTS schema_migrations (
      id         SERIAL       PRIMARY KEY,
      filename   TEXT         NOT NULL UNIQUE,
      applied_at TIMESTAMPTZ  NOT NULL DEFAULT NOW()
    )
  `);

  const migrationsDir = path.join(__dirname, 'migrations');
  const files = fs
    .readdirSync(migrationsDir)
    .filter((f) => f.endsWith('.sql'))
    .sort();

  for (const file of files) {
    const already = await pool.query(
      'SELECT 1 FROM schema_migrations WHERE filename = $1',
      [file]
    );
    if (already.rows.length > 0) {
      console.log(`Skipping ${file} (already applied)`);
      continue;
    }

    console.log(`Applying ${file}...`);
    const sql = fs.readFileSync(path.join(migrationsDir, file), 'utf8');
    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      await client.query(sql);
      await client.query(
        'INSERT INTO schema_migrations (filename) VALUES ($1)',
        [file]
      );
      await client.query('COMMIT');
      console.log(`  ✓ ${file}`);
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  }

  await pool.end();
}

migrate().catch((err) => {
  console.error('Migration failed:', err);
  process.exit(1);
});

File: backend/db/migrate.py
import os
import pathlib

from backend.src.db.client import get_connection


def _bootstrap_migrations_table(conn: object) -> None:
    import psycopg2.extensions
    c: psycopg2.extensions.connection = conn  # type: ignore[assignment]
    with c.cursor() as cur:
        cur.execute(
            """
            CREATE TABLE IF NOT EXISTS schema_migrations (
                filename TEXT PRIMARY KEY,
                applied_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
            )
            """
        )
    c.commit()


def _applied_migrations(conn: object) -> set[str]:
    import psycopg2.extensions
    c: psycopg2.extensions.connection = conn  # type: ignore[assignment]
    with c.cursor() as cur:
        cur.execute("SELECT filename FROM schema_migrations")
        return {row[0] for row in cur.fetchall()}


def run_migrations() -> None:
    migrations_dir = pathlib.Path(__file__).parent / "migrations"
    sql_files = sorted(migrations_dir.glob("*.sql"))

    conn = get_connection()
    try:
        _bootstrap_migrations_table(conn)
        applied = _applied_migrations(conn)
        for sql_file in sql_files:
            if sql_file.name in applied:
                print(f"Skipping {sql_file.name} (already applied)")
                continue
            print(f"Applying {sql_file.name}...")
            sql = sql_file.read_text()
            with conn.cursor() as cur:
                cur.execute(sql)
                cur.execute(
                    "INSERT INTO schema_migrations (filename) VALUES (%s)",
                    (sql_file.name,),
                )
            conn.commit()
            print(f"Applied {sql_file.name}")
    finally:
        conn.close()


def main() -> None:
    run_migrations()
    print("Migrations complete.")


if __name__ == "__main__":
    main()
